packages/e2e-test-utils-playwright/src/metrics/index.ts

```diff
@@ -273,4 +289,95 @@ export class Metrics {
 			)
 			.map( ( item ) => ( item.dur ? item.dur / 1000 : 0 ) );
 	}
+
+	/**
+	 * Initializes the web-vitals library upon next page navigation.
+	 *
+	 * Defaults to automatically triggering the navigation,
+	 * but it can also be done manually.
+	 *
+	 * @example
+	 * ```js
+	 * await metrics.initWebVitals();
+	 * console.log( await metrics.getWebVitals() );
+	 * ```
+	 *
+	 * @example
+	 * ```js
+	 * await metrics.initWebVitals( false );
+	 * await page.goto( '/some-other-page' );
+	 * console.log( await metrics.getWebVitals() );
+	 * ```
+	 *
+	 * @param reload Whether to force navigation by reloading the current page.
+	 */
+	async initWebVitals( reload = true ) {
+		await this.page.addInitScript( {
+			path: join(
+				__dirname,
+				'../../../../node_modules/web-vitals/dist/web-vitals.umd.cjs'
+			),
+		} );
+
+		await this.page.exposeFunction(
+			'__reportVitals__',
+			( data: string ) => {
+				const measurement: Metric = JSON.parse( data );
+				this.webVitals[ measurement.name ] = measurement.value;
+			}
+		);
+
+		await this.page.addInitScript( () => {
+			const reportVitals = ( measurement: unknown ) =>
+				window.__reportVitals__( JSON.stringify( measurement ) );
+
+			window.addEventListener( 'DOMContentLoaded', () => {
+				// @ts-ignore
+				window.webVitals.onCLS( reportVitals );
+				// @ts-ignore
+				window.webVitals.onFCP( reportVitals );
+				// @ts-ignore
+				window.webVitals.onFID( reportVitals );
+				// @ts-ignore
+				window.webVitals.onINP( reportVitals );
+				// @ts-ignore
+				window.webVitals.onLCP( reportVitals );
+				// @ts-ignore
+				window.webVitals.onTTFB( reportVitals );
+			} );
```

Comment:
Have you verified there is no issue in using `DOMContentLoaded` here? Couldn't any of these events fires before `DOMContentLoaded` (thinking TTFB for example)?

Related is my above question, I don't think we would need to do this if we injected a single `type="module"` script that imports the functions from `web-vitals` and then calls them directly rather than depending on the `window` global.

Reply:
They are not really events per se, rather functions for you to add your own callbacks. You call it once per page load and then once the metrics are reported (on page unload or interaction), your callback is invoked.

The only issue I could think of is if you are not awaiting for the page to load when doing some interaction (i.e. you click too early on something). But doing `await page.goto()` or `page.reload()` in Playwright auto-waits for the `load` event automatically (which fires after `DOMContentLoaded`). And disabling auto-awaiting in PW is highloy discouraged anyway, so chances of that happening are really low.

I also verified expected behavior by running this locally. Numbers are reported just fine.

Reply:
Right, I meant what if one of these metrics is reported before `DOMContentLoaded`? Couldn't the TTFB reporting fire before that (since that metric is at least in principle _available_ before)?

Why do we need to use `DOMContentLoaded` to add our callbacks to these functions? Shouldn't `window.webVitals` already be available before?

Reply:
I'm 99% sure that the reporting is `buffered`, so even if the event already transpired before invoking here, the callback will still fire.

Reply:
Yes, they‘re buffered.

Can indeed probably omit the DOMContentLoaded part as I think the script is added inline by PW, but will double check.

Reply:
Actually, I cannot remove the `DOMContentLoaded` part, as the order of evaluation of multiple scripts installed via  [page.addInitScript()](https://playwright.dev/docs/api/class-page#page-add-init-script) is not defined ([docs](https://playwright.dev/docs/api/class-page#page-add-init-script)).

So `DOMContentLoaded` here is just a simple workaround to wait for the first script to have successfully loaded.

Again, I don't see an issue with using this event listener here.

And everything is buffered as well, so you don't miss anything.

If you test the script yourself you'll see that it works just fine.

Reply:
Thanks for clarifying. All good then.